# Model Evaluation

Generated from `notebooks/DELE_CA1_B.ipynb` by `scripts/split_notebook.py`.


---
# 9. Model Evaluation

In this section, we will be evaluating all the Model's performance for both Classification and Regression and for both Data with Augmentation and No Augmentation. This is to compare and evaluate the various RNN Models trained and determine which Model will be the most suitable to be used.

---
## 9.1 Classification Model Evaluation

In this sub-section, we will be evaluating the Classification RNN Models and compare the Test Accuracy and Lost to one another. This will help us to determine the best RNN Classification Model and also to evaluate whether Data Augmentation assisted in the Training of the RNN Model and Improvement of the Accuracy Metrics.

In [ ]:
# ==================== Classification Model Paths ==================== #
classification_model_paths = {
    "SimpleRNN Classification (Aug)": "Models/srnn_classification.keras",
    "SimpleRNN Classification (No Aug)": "Models/srnn_classification_no_aug.keras",
    "LSTM Classification (Aug)": "Models/lstm_classification.keras",
    "LSTM Classification (No Aug)": "Models/lstm_classification_no_aug.keras",
    "GRU Classification (Aug)": "Models/gru_classification.keras",
    "GRU Classification (No Aug)": "Models/gru_classification_no_aug.keras"
}

# ==================== Evaluate Classification Models ==================== #
classification_results = []

for model_name, path in classification_model_paths.items():
    try:
        model = load_model(path)
        loss, acc = model.evaluate(X_test_class_pad, y_test_class, verbose=0)
        is_augmented = "(Aug)" in model_name

        classification_results.append({
            "Model": model_name,
            "Test Accuracy": round(acc, 4),
            "Test Loss": round(loss, 4),
            "Augmented": "Yes" if is_augmented else "No"
        })
    except Exception as e:
        classification_results.append({
            "Model": model_name,
            "Test Accuracy": None,
            "Test Loss": None,
            "Augmented": "Unknown",
            "Error": str(e)
        })

# ==================== Format and Display ==================== #
df_classification_results = pd.DataFrame(classification_results)
df_classification_results = df_classification_results.sort_values(by="Test Accuracy", ascending=False).reset_index(drop=True)
df_classification_results.style.background_gradient(cmap="Blues")

,Model,Test Accuracy,Test Loss,Augmented
0,GRU Classification (Aug),0.964300,1.055000,Yes
1,SimpleRNN Classification (Aug),0.750000,1.357200,Yes
2,LSTM Classification (No Aug),0.750000,1.693700,No
3,SimpleRNN Classification (No Aug),0.678600,1.794300,No
4,LSTM Classification (Aug),0.678600,2.132200,Yes
5,GRU Classification (No Aug),0.383100,0.625400,No


With reference to the output of the DataFrame above, we are able to make the following observations:

**GRU Outperformed All Others:**

- The GRU Classification model with augmentation achieved the highest accuracy (0.9286).

- This supports the hypothesis that data augmentation boosts GRU's ability to generalise.

**SimpleRNN Without Augmentation Is Surprisingly Strong:**

- Despite not using augmented data, SimpleRNN achieved second-highest accuracy (0.8929).

- This suggests that SimpleRNN might already generalise well on smaller datasets, or possibly overfits less.

**Mixed Results for Augmentation:**

- While augmentation helped GRU, it did not help LSTM, which still had lower accuracy whether augmented (0.5357) or not (0.4343).

- This could imply LSTM requires more tuning, or is more sensitive to noisy or less relevant augmentation.

**Model Stability Matters:**

- The Test Loss for LSTM (Aug) was very high (3.6299), suggesting instability or poor convergence.

- Conversely, GRU (Aug) maintained both high accuracy and a reasonable loss, showing robustness to augmentation.

With the observations stated, we will now move on to the Regression Models' Evaluation in the next sub-section

---
## 9.2 Regression Model Evaluation

In this sub-section, we will be evaluating the Regression RNN Models and compare the MAPE to one another. This will help us to determine the best RNN Regression Model and also to evaluate whether Data Augmentation assisted in the Training of the RNN Model and Improvement of the Accuracy Metrics.

In [ ]:
# ==================== Regression Model Paths ==================== #
regression_model_paths = {
    "SimpleRNN Regression (Aug)": "Models/srnn_regression.keras",
    "SimpleRNN Regression (No Aug)": "Models/srnn_regression_no_aug.keras",
    "LSTM Regression (Aug)": "Models/lstm_regression.keras",
    "LSTM Regression (No Aug)": "Models/lstm_regression_no_aug.keras",
    "GRU Regression (Aug)": "Models/gru_regression.keras",
    "GRU Regression (No Aug)": "Models/gru_regression_no_aug.keras"
}

# ==================== Evaluate Regression Models ==================== #
regression_results = []

for model_name, path in regression_model_paths.items():
    try:
        model = load_model(path)
        y_pred = model.predict(X_test_reg_pad).flatten()
        mape = mean_absolute_percentage_error(y_test_reg, y_pred)
        is_augmented = "(Aug)" in model_name

        regression_results.append({
            "Model": model_name,
            "MAPE": round(mape, 4),
            "Augmented": "Yes" if is_augmented else "No"
        })
    except Exception as e:
        regression_results.append({
            "Model": model_name,
            "MAPE": None,
            "Augmented": "Unknown",
            "Error": str(e)
        })

# ==================== Format and Display ==================== #
df_regression_results = pd.DataFrame(regression_results)
df_regression_results = df_regression_results.sort_values(by="MAPE", ascending=True).reset_index(drop=True)
df_regression_results.style.background_gradient(cmap="Blues")

1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 529ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 1s 529ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 280ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 258ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 239ms/step
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 237ms/step


,Model,MAPE,Augmented
0,GRU Regression (Aug),0.499300,Yes
1,GRU Regression (No Aug),0.535600,No
2,SimpleRNN Regression (Aug),0.571500,Yes
3,LSTM Regression (Aug),0.584100,Yes
4,SimpleRNN Regression (No Aug),0.690500,No
5,LSTM Regression (No Aug),0.698400,No


With reference to the Code Cell above, we are able to make the following Observations:

**LSTM (No Aug) Is the Best Performing Model:**

- Achieved the lowest MAPE (0.5114), indicating the most accurate predictions relative to the true values.

- Suggests that LSTM benefits more from raw data quality than quantity expansion via augmentation.

**GRU Models Show Consistent Stability:**

- Both GRU with and without augmentation performed relatively well (0.5588 vs 0.5648).

- This reflects GRU’s ability to retain learning performance under varied data conditions, showing robustness.

**Augmentation Didn't Always Help:**

- Augmented versions of LSTM and SimpleRNN had worse MAPE than their non-augmented counterparts.

- Implies that in regression, especially with nuanced sentiment scores, augmentation may introduce semantic drift, worsening model performance.

**SimpleRNN Struggles in Regression:**

- It had the highest MAPE regardless of augmentation (0.6838 no aug, 0.5715 with aug).

- Suggests that SimpleRNN lacks the depth or temporal precision needed for high-quality continuous value predictions.

With the observations indicated, we will now move on to selecting the Method and the Best Model to conduct further evaluation in the next section.